ipc_sendfile.c

```diff
@@ -0,0 +1,203 @@
+		{ "file",     1, NULL, 'f' },
+		{ NULL,       0, NULL, 0   }   /* Required at end of array.  */
+	};
+
+	int next_option_fileName;
+	do { // loop for finding the fileName
+		next_option_fileName = getopt_long (argc, argv, short_options,long_options, NULL);
+		switch (next_option_fileName){
+		case 'f': /* -f or --file */
+			/* This option takes an argument, the name of the output file.  */
+		    fileName = optarg;
+		    break;
+		}
+	}
+	while (next_option_fileName != -1 && fileName == NULL);
+
+	optind=0; // to reset the options index for our new loop
+
+	do {
+	    next_option = getopt_long (argc, argv, short_options,long_options, NULL);
+	    switch (next_option)
+	    {
+	    case 'h':   /* -h or --help */
+	    	display_help();
+	    	break;
+
+	    case 'm':   /* -m or --messages */
+	    	if (fileName == NULL){
+	    		printf("ERROR: To send an IPC message type:\nipc_sendfile --messages --file <path_to_source/file>\nfor more informations:   ipc_sendfile --help    \n");
+	    		exit(EXIT_FAILURE);
+	    	}
+	    	return ipc_send_message(fileName);
+
+	    case '?':   /* The user specified an invalid option.  */
+	    	display_arg_error();
+	    	exit(EXIT_FAILURE);
+
+	    default:    /* Something else: unexpected.  */
+	    	exit(EXIT_FAILURE);
+	    }
+	  }
+	  while (next_option != -1);
+
+   return 0;
+}
+void display_help(void)
+{
+    printf("\nThis program is made to send a file to ipc_receivefile \n");
+    printf("Commands:\n");
+    printf("To display this message: ipc_sendfile --help \n");
+    printf("To send a file using messages: ipc_sendfile --messages --file <path_to_source/file> \n");
+}
+void display_arg_error(void)
+{
+    printf("ERROR: This program must be started with commandline arguments, for example:\n\n");
+	printf("   ipc_sendfile --help    \n\n");
+}
+int ipc_send_message(char* fileName){
+
+	printf("Preparing to receive a message:\n");
+
+	int coid; //Connection ID to server
+	int status;
+	int incoming_status;
+	int fileSize;
+
+	coid =-1;
+
+	while (coid == -1){
+		printf("Connecting to message receiver... \n");
+		coid = name_open(RECV_NAME, 0);
+		if (coid == -1){
+			sleep(2);
+		}
+	}
+	printf("Connected to the receiver \n");
+
+	fileSize=filesize(fileName);
+
+	int i;
+
+	printf("file size: %d\n",fileSize);
+
+	msg_header_t header;
+	header.data_size=fileSize;
+	header.msg_type=IOV_MSG_TYPE;
+
+	int last_part_size = fileSize%iov_block_size;
+	printf("iov nb = %d\n",(fileSize/iov_block_size)+2);
+	iov_t siov[(fileSize/iov_block_size)+2];
+
+    unsigned char* buffer=readFile(fileName); //getting the data into the buffer
+
+	SETIOV (&siov[0], &header, sizeof header); //setting header IOV
+
+	for(i=1;i<=fileSize/iov_block_size+1;i++){ //setting data IOVs
+		if (i == fileSize/iov_block_size+1){ //in case it is the last IOV
+			SETIOV (&siov[i], &buffer[(i-1)*iov_block_size], last_part_size);
+		} else { //else
+			SETIOV (&siov[i], &buffer[(i-1)*iov_block_size], iov_block_size);
+		}
```

Comment:
some comments explaining the processes can be helpful too.

Reply:
Done!